File: src/petcast/dither.py
"""Spectra 6 e-ink dithering pipeline.

Matches the ESPHome epaper_spi Spectra E6 driver's color classification:
grayscale check (spread < 50), then binary threshold at 128 per channel.
"""

import numpy as np
from PIL import Image, ImageEnhance

from petcast.config import Config

# Output palette values — one per driver bucket
# These must pass through the driver's classification correctly
SPECTRA6_PALETTE = {
    "BLACK": (0, 0, 0),
    "WHITE": (255, 255, 255),
    "RED": (200, 0, 0),
    "GREEN": (0, 150, 0),
    "BLUE": (0, 0, 200),
    "YELLOW": (255, 230, 0),
}


def _driver_classify(r: float, g: float, b: float) -> str:
    """Replicate the ESPHome Spectra E6 driver's color classification."""
    ri, gi, bi = int(np.clip(r, 0, 255)), int(np.clip(g, 0, 255)), int(np.clip(b, 0, 255))
    spread = max(ri, gi, bi) - min(ri, gi, bi)

    if spread < 50:
        # Grayscale
        return "WHITE" if (ri + gi + bi) > 382 else "BLACK"

    # Binary threshold per channel
    ro = ri > 128
    go = gi > 128
    bo = bi > 128

    if ro and go and not bo:
        return "YELLOW"
    if ro and not go and not bo:
        return "RED"
    if not ro and go and not bo:
        return "GREEN"
    if not ro and not go and bo:
        return "BLUE"
    if not ro and go and bo:
        return "GREEN"  # cyan → green
    if ro and not go and bo:
        return "RED"    # magenta → red
    if ro and go and bo:
        return "WHITE"
    return "BLACK"


def dither_for_display(image: Image.Image, config: Config) -> Image.Image:
    """Resize, enhance, and dither an image for Spectra 6 e-ink display."""
    w, h = config.display.width, config.display.height

    img = image.convert("RGB")
    img = _resize_crop(img, w, h)

    # Boost contrast and saturation for e-ink
    img = ImageEnhance.Contrast(img).enhance(1.2)
    img = ImageEnhance.Color(img).enhance(1.3)

    img = _floyd_steinberg_dither(img)

    return img


def _resize_crop(img: Image.Image, target_w: int, target_h: int) -> Image.Image:
    """Resize and center-crop to exactly target dimensions."""
    src_w, src_h = img.size
    target_ratio = target_w / target_h
    src_ratio = src_w / src_h

    if src_ratio > target_ratio:
        new_h = target_h
        new_w = int(src_w * (target_h / src_h))
    else:
        new_w = target_w
        new_h = int(src_h * (target_w / src_w))

    img = img.resize((new_w, new_h), Image.Resampling.LANCZOS)

    left = (new_w - target_w) // 2
    top = (new_h - target_h) // 2
    return img.crop((left, top, left + target_w, top + target_h))


def _floyd_steinberg_dither(img: Image.Image) -> Image.Image:
    """Floyd-Steinberg dithering using the driver's own color classification."""
    pixels = np.array(img, dtype=np.float64)
    h, w, _ = pixels.shape
    palette = SPECTRA6_PALETTE

    for y in range(h):
        for x in range(w):
            old = pixels[y, x].copy()

            # Classify using the same logic as the display driver
            color_name = _driver_classify(old[0], old[1], old[2])
            new = np.array(palette[color_name], dtype=np.float64)

            pixels[y, x] = new
            error = old - new

            # Distribute error to neighbors
            if x + 1 < w:
                pixels[y, x + 1] += error * (7 / 16)
            if y + 1 < h:
                if x - 1 >= 0:
                    pixels[y + 1, x - 1] += error * (3 / 16)
                pixels[y + 1, x] += error * (5 / 16)
                if x + 1 < w:
                    pixels[y + 1, x + 1] += error * (1 / 16)

    pixels = np.clip(pixels, 0, 255).astype(np.uint8)
    return Image.fromarray(pixels)

File: internal/pipeline/dither.go
package pipeline

import (
	"image"
	"image/color"
	"math"

	"golang.org/x/image/draw"
)

// Spectra6Palette is the 6-color palette for Spectra 6 e-ink displays.
var Spectra6Palette = []color.NRGBA{
	{0, 0, 0, 255},       // Black
	{255, 255, 255, 255},  // White
	{200, 0, 0, 255},      // Red
	{0, 150, 0, 255},      // Green
	{0, 0, 200, 255},      // Blue
	{255, 230, 0, 255},    // Yellow
}

// DitherForDisplay resizes, enhances, and applies Atkinson dithering for e-ink.
func DitherForDisplay(src image.Image, targetW, targetH int) *image.NRGBA {
	// Resize and center-crop
	resized := resizeCrop(src, targetW, targetH)

	// Boost contrast (1.2) and saturation (1.3) for e-ink
	enhanced := enhanceContrast(resized, 1.2)
	enhanced = enhanceSaturation(enhanced, 1.3)

	// Atkinson dithering
	return atkinsonDither(enhanced, Spectra6Palette)
}

func resizeCrop(src image.Image, targetW, targetH int) *image.NRGBA {
	srcBounds := src.Bounds()
	srcW := srcBounds.Dx()
	srcH := srcBounds.Dy()

	targetRatio := float64(targetW) / float64(targetH)
	srcRatio := float64(srcW) / float64(srcH)

	var newW, newH int
	if srcRatio > targetRatio {
		newH = targetH
		newW = int(float64(srcW) * (float64(targetH) / float64(srcH)))
	} else {
		newW = targetW
		newH = int(float64(srcH) * (float64(targetW) / float64(srcW)))
	}

	// Resize with CatmullRom (high quality)
	scaled := image.NewNRGBA(image.Rect(0, 0, newW, newH))
	draw.CatmullRom.Scale(scaled, scaled.Bounds(), src, srcBounds, draw.Over, nil)

	// Center crop
	left := (newW - targetW) / 2
	top := (newH - targetH) / 2
	cropped := image.NewNRGBA(image.Rect(0, 0, targetW, targetH))
	draw.Copy(cropped, image.Point{}, scaled, image.Rect(left, top, left+targetW, top+targetH), draw.Src, nil)
	return cropped
}

func enhanceContrast(img *image.NRGBA, factor float64) *image.NRGBA {
	bounds := img.Bounds()
	out := image.NewNRGBA(bounds)
	for y := bounds.Min.Y; y < bounds.Max.Y; y++ {
		for x := bounds.Min.X; x < bounds.Max.X; x++ {
			c := img.NRGBAAt(x, y)
			r := clampU8(128 + factor*(float64(c.R)-128))
			g := clampU8(128 + factor*(float64(c.G)-128))
			b := clampU8(128 + factor*(float64(c.B)-128))
			out.SetNRGBA(x, y, color.NRGBA{r, g, b, c.A})
		}
	}
	return out
}

func enhanceSaturation(img *image.NRGBA, factor float64) *image.NRGBA {
	bounds := img.Bounds()
	out := image.NewNRGBA(bounds)
	for y := bounds.Min.Y; y < bounds.Max.Y; y++ {
		for x := bounds.Min.X; x < bounds.Max.X; x++ {
			c := img.NRGBAAt(x, y)
			gray := 0.2989*float64(c.R) + 0.5870*float64(c.G) + 0.1140*float64(c.B)
			r := clampU8(gray + factor*(float64(c.R)-gray))
			g := clampU8(gray + factor*(float64(c.G)-gray))
			b := clampU8(gray + factor*(float64(c.B)-gray))
			out.SetNRGBA(x, y, color.NRGBA{r, g, b, c.A})
		}
	}
	return out
}

// atkinsonDither applies Atkinson dithering with the given palette.
// Diffuses 1/8 of error to 6 neighbors (75% total error diffusion).
func atkinsonDither(img *image.NRGBA, palette []color.NRGBA) *image.NRGBA {
	bounds := img.Bounds()
	w := bounds.Dx()
	h := bounds.Dy()

	// Work with float64 pixel buffer for error diffusion
	pixels := make([]float64, w*h*3)
	for y := 0; y < h; y++ {
		for x := 0; x < w; x++ {
			c := img.NRGBAAt(x+bounds.Min.X, y+bounds.Min.Y)
			idx := (y*w + x) * 3
			pixels[idx] = float64(c.R)
			pixels[idx+1] = float64(c.G)
			pixels[idx+2] = float64(c.B)
		}
	}

	// Pre-convert palette to float64
	pal := make([][3]float64, len(palette))
	for i, c := range palette {
		pal[i] = [3]float64{float64(c.R), float64(c.G), float64(c.B)}
	}

	out := image.NewNRGBA(image.Rect(0, 0, w, h))

	for y := 0; y < h; y++ {
		for x := 0; x < w; x++ {
			idx := (y*w + x) * 3
			oldR := pixels[idx]
			oldG := pixels[idx+1]
			oldB := pixels[idx+2]

			// Find nearest palette color
			best := 0
			bestDist := math.MaxFloat64
			for i, p := range pal {
				dr := oldR - p[0]
				dg := oldG - p[1]
				db := oldB - p[2]
				dist := dr*dr + dg*dg + db*db
				if dist < bestDist {
					bestDist = dist
					best = i
				}
			}

			newR := pal[best][0]
			newG := pal[best][1]
			newB := pal[best][2]

			out.SetNRGBA(x, y, palette[best])

			// Error = (old - new) / 8
			errR := (oldR - newR) / 8.0
			errG := (oldG - newG) / 8.0
			errB := (oldB - newB) / 8.0

			// Diffuse to 6 neighbors (Atkinson pattern)
			diffuse := func(dx, dy int) {
				nx, ny := x+dx, y+dy
				if nx >= 0 && nx < w && ny >= 0 && ny < h {
					nIdx := (ny*w + nx) * 3
					pixels[nIdx] += errR
					pixels[nIdx+1] += errG
					pixels[nIdx+2] += errB
				}
			}
			diffuse(1, 0)
			diffuse(2, 0)
			diffuse(-1, 1)
			diffuse(0, 1)
			diffuse(1, 1)
			diffuse(0, 2)
		}
	}

	return out
}

func clampU8(v float64) uint8 {
	if v < 0 {
		return 0
	}
	if v > 255 {
		return 255
	}
	return uint8(v + 0.5)
}
